src/lem/store.rs

```diff
@@ -721,88 +847,67 @@ impl<F: LurkField> Store<F> {
     /// Warning: without cache hits, this function might blow up Rust's recursion
     /// depth limit. This limitation is circumvented by calling `hydrate_z_cache`
     /// beforehand or by using `hash_ptr` instead, which is slightly slower.
-    fn hash_ptr_unsafe(&self, ptr: &Ptr) -> ZPtr<F> {
+    fn hash_raw_ptr_unsafe(&self, ptr: &RawPtr) -> FWrap<F> {
         match ptr {
-            Ptr::Atom(tag, idx) => {
-                if let Some(z_ptr) = self.z_cache.get(ptr) {
-                    *z_ptr
+            RawPtr::Atom(idx) => FWrap(*self.expect_f(*idx)),
+            RawPtr::Hash3(idx) => {
+                if let Some(z) = self.z_cache.get(ptr) {
+                    *z
                 } else {
-                    let z_ptr = ZPtr::from_parts(*tag, *self.expect_f(*idx));
-                    self.z_cache.insert(*ptr, Box::new(z_ptr));
-                    self.inverse_z_cache.insert(z_ptr, Box::new(*ptr));
-                    z_ptr
+                    let children_ptrs = self.expect_raw_ptrs::<3>(*idx);
+                    let mut children_zs = [F::ZERO; 3];
+                    for (idx, child_ptr) in children_ptrs.iter().enumerate() {
+                        children_zs[idx] = self.hash_raw_ptr_unsafe(child_ptr).0;
+                    }
+                    let z = FWrap(self.poseidon_cache.hash3(&children_zs));
+                    self.z_cache.insert(*ptr, Box::new(z));
+                    self.inverse_z_cache.insert(z, Box::new(*ptr));
+                    z
```

Comment:
I think these lines can be factored out to a single function (they repeat for all kinds of `RawPtr::Hash*`)

Also, would we benefit from calling [`par_iter_mut`](https://docs.rs/rayon/latest/rayon/iter/trait.IntoParallelRefMutIterator.html#tymethod.par_iter_mut) on `children_ptrs`?